packages/compiler-cli/src/ngtsc/annotations/src/diagnostics.ts

```diff
@@ -0,0 +1,114 @@
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import * as ts from 'typescript';
+
+import {ErrorCode, makeDiagnostic} from '../../diagnostics';
+import {Reference} from '../../imports';
+import {MetadataReader} from '../../metadata';
+import {PartialEvaluator} from '../../partial_evaluator';
+import {ClassDeclaration, ReflectionHost, isNamedClassDeclaration} from '../../reflection';
+
+import {readBaseClass} from './util';
+
+export function getDirectiveDiagnostics(
+    node: ClassDeclaration, reader: MetadataReader, reflector: ReflectionHost,
+    evaluator: PartialEvaluator): ts.Diagnostic[]|null {
+  let diagnostics: ts.Diagnostic[]|null = [];
+
+  const addDiagnostics = (more: ts.Diagnostic | ts.Diagnostic[] | null) => {
+    if (more === null) {
+      return;
+    } else if (diagnostics === null) {
+      diagnostics = Array.isArray(more) ? more : [more];
+    } else if (Array.isArray(more)) {
+      diagnostics.push(...more);
+    } else {
+      diagnostics.push(more);
+    }
+  };
+
+  addDiagnostics(checkInheritanceOfDirective(node, reader, reflector, evaluator));
+  return diagnostics;
+}
+
+export function checkInheritanceOfDirective(
+    node: ClassDeclaration, reader: MetadataReader, reflector: ReflectionHost,
+    evaluator: PartialEvaluator): ts.Diagnostic|null {
+  if (!ts.isClassDeclaration(node)) {
+    return null;
+  }
+
+  if (getConstructor(node) !== undefined) {
+    // If a constructor exists, then no base class definition is required on the runtime side -
+    // it's legal to inherit from any class.
+    return null;
+  }
+
+  // The extends clause is an expression which can be as dynamic as the user wants. Try to
+  // evaluate it, but fall back on ignoring the clause if it can't be understood. This is a View
+  // Engine compatibility hack: View Engine ignores 'extends' expressions that it cannot understand.
+  let baseClass = readBaseClass(node, reflector, evaluator);
+
+  while (baseClass) {
+    if (baseClass === 'dynamic' || !isNamedClassDeclaration(baseClass.node)) {
+      return null;
+    }
+
+    // We can skip the base class if it has metadata.
+    const baseClassMeta = reader.getDirectiveMetadata(baseClass);
+    if (baseClassMeta !== null) {
+      return null;
+    }
+
+    // If the base class has a blank constructor we can skip it since it can't be using DI.
+    const baseClassConstructor = getConstructor(baseClass.node);
+    if (baseClassConstructor) {
+      if (baseClassConstructor.parameters.length === 0) {
+        return null;
+      } else {
+        return getInheritedUndecoratedCtorDiagnostic(node, baseClass, reader);
+      }
+    }
+
+    const newParentClass = readBaseClass(baseClass.node, reflector, evaluator);
+    if (newParentClass !== null) {
+      // If we found another parent class, keep going.
+      baseClass = newParentClass;
+    } else if (!baseClassConstructor) {
```

Comment:
In the compiler, prefer explicit comparisons and don't rely on type coercion:

```suggestion
    } else if (baseClassConstructor === null) {
```